File: IT/Strutture dati/Esercizi/C++/esercizio29.cpp
/*
file esercizio29.cpp
compilare con: g++ -std=c++11 esercizio29.cpp
*/

#include <iostream>
#include <iterator>
#include <random>
#include <algorithm>

using std::cout;
using std::endl;
using std::ostream_iterator;
using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::sort;

int main()
{
	default_random_engine gen(random_device{}());
	uniform_int_distribution<int> rrange(1, 20);

	const int LEN = rrange(gen) * 5;
	int *nums = new int[LEN];

	uniform_int_distribution<int> dist(1, 100);

	for (int i = 0; i < LEN; i++) {
		nums[i] = dist(gen);
	}

	sort(nums, nums+LEN);

	cout << "[";
	copy(nums, nums+LEN, ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	int md = 1;
	int counts[LEN];
	int i = 0;

	while (i < LEN) {
		int n = nums[i];
		int j = i+1; // indice dell'elemento successivo
		int occ = 1;
		while (j < LEN && n == nums[j]) {
			occ++;
			j++;
		}
		counts[i] = occ; // memorizza il numero di occorrenze
		if (occ > md) md = occ;
		i += occ; // salta al prossimo numero
	}

	if (md == 1) {
		cout << "L'array non ha una moda." << endl;
		return 0;
	}

	cout << "Moda (" << md << " occorrenze): ";

	int j = 0;
	for (int i = 0; i < LEN; i++) {
		if (counts[i] == md) {
			if (j > 0) cout << ", ";
			cout << nums[i];
			j++;
		}
	}

	cout << "." << endl;

	delete [] nums;
}

File: IT/Strutture dati/Esercizi/Java/Esercizio29.java
/* file Esercizio29.java */

import java.util.Arrays;
import java.util.Random;

public class Esercizio29 {
	public static void main(String [] args) {
		Random rand = new Random();
		final int LEN = rand.nextInt(20) + 1;
		int [] nums = new int[LEN*5];

		for (int i = 0; i < nums.length; i++) {
			nums[i] = rand.nextInt(100) + 1;
		}

		System.out.println(Arrays.toString(nums));

		Arrays.sort(nums);

		System.out.println(Arrays.toString(nums));

		int md = 1;
		int [] counts = new int[nums.length];
		int i = 0;

		while (i < nums.length) {
			int n = nums[i];
			int j = i+1; // indice dell'elemento successivo
			int occ = 1;
			while (j < nums.length && n == nums[j]) {
				occ++;
				j++;
			}
			counts[i] = occ; // memorizza il numero di occorrenze
			if (occ > md) md = occ;
			i += occ; // salta al prossimo numero
		}

		if (md == 1) {
			System.out.println("L'array non presenta una moda.");
			return;
		}

		System.out.print("Moda (" + md + " occorrenze): ");

		int k = 0;
		for (int j = 0; j < counts.length; j++) {
			if (counts[j] == md) {
				if (k > 0) System.out.print(", ");
				System.out.print(nums[j]);
				k++;
			}
		}

		System.out.println(".");
	}
}

